internal/controllers/configuration/secret_controller.go

```diff
@@ -0,0 +1,134 @@
+package configuration
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	ctrlref "github.com/kong/kubernetes-ingress-controller/v2/internal/controllers/reference"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/util"
+)
+
+// -----------------------------------------------------------------------------
+// CoreV1 Secret - Reconciler
+// -----------------------------------------------------------------------------
+
+const (
+	CACertLabelKey = "konghq.com/ca-cert"
+)
+
+// CoreV1SecretReconciler reconciles Secret resources
+type CoreV1SecretReconciler struct {
+	client.Client
+
+	Log              logr.Logger
+	Scheme           *runtime.Scheme
+	DataplaneClient  *dataplane.KongClient
+	CacheSyncTimeout time.Duration
+
+	ReferenceIndexers ctrlref.CacheIndexers
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *CoreV1SecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c, err := controller.New("CoreV1Secret", mgr, controller.Options{
+		Reconciler: r,
+		LogConstructor: func(_ *reconcile.Request) logr.Logger {
+			return r.Log
+		},
+		CacheSyncTimeout: r.CacheSyncTimeout,
+	})
+	if err != nil {
+		return err
+	}
+
+	predicateFuncs := predicate.NewPredicateFuncs(r.shouldReconcileSecret)
+	//we should always try to delete secrets in caches when they are deleted in cluster.
+	predicateFuncs.DeleteFunc = func(event event.DeleteEvent) bool { return true }
+	return c.Watch(
+		&source.Kind{Type: &corev1.Secret{}},
+		&handler.EnqueueRequestForObject{},
+		predicateFuncs,
+	)
+}
+
+// shouldReconcileSecret is the filter function to judge whether the secret should be reconciled
+// and stored in cache of the controller. It returns true for the secret should be reconciled when:
+// - the secret has label: konghq.com/ca-cert:true
+// - or the secret is referred by objects we care (service, ingress, gateway, ...)
+func (r *CoreV1SecretReconciler) shouldReconcileSecret(obj client.Object) bool {
+	secret, ok := obj.(*corev1.Secret)
+	if !ok {
+		return false
+	}
+
+	labels := secret.Labels
+	if labels != nil && labels[CACertLabelKey] == "true" {
+		return true
+	}
+
+	referred, err := r.ReferenceIndexers.ObjectReferred(secret)
+
+	if err != nil {
+		r.Log.Info("failed to check whether secret referred:", err)
```

Comment:
How often can this actually happen 🤔 ? Is `.Info` a good level for this log entry? Maybe a warning/error would be more suitable (if this doesn't happen too often)?

Reply:
Changed to `Error` level and added more information (namespace/name) in log.